### PR TITLE
[Addressing] Prevent deleting zones and provinces that are zone members

### PR DESCRIPTION
**Pull request by coldic3**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.11
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no
| Related tickets |                      |
| License         | MIT                                                          |

It is possible to delete the resource (zone or province) that is already in use as a zone member. When we delete such a resource, we cannot access the zone edit screen (and probably even more) because we try to load an unexisting resource 💥

Sample error:
```
Unable to transform data for property path "code": Object "Sylius\Component\Addressing\Model\Zone" with identifier "code"="LETS_DELETE_ME" does not exist.
```

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Admin/ManagingCountriesContext.php [hunk 1]
```diff
@@ -348,6 +348,17 @@ public function iShouldBeNotifiedThatNameOfTheProvinceIsRequired(): void
...
     }
 
+    /**
+     * @Then I should be notified that provinces that are in use cannot be deleted
+     */
+    public function iShouldBeNotifiedThatProvincesThatAreInUseCannotBeDeleted(): void
+    {
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            'Cannot delete, the province is in use.'
```
Comment (Ferror) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCountriesContext.php:358]: This error message and `Error Cannot delete, the province is in use.` src/Sylius/Behat/Context/Ui/Admin/ManagingCountriesContext.php:284 shouldn't be more likely the same?

Comment (coldic3) [reply src/Sylius/Behat/Context/Api/Admin/ManagingCountriesContext.php:358]: "Error" before the error message is something that exists only in flash messages as a message header.
image

I think it's a good idea to compare just a message in the UI context but I would be for doing it in separate PR as it requires changes in many behat contexts.

File: features/addressing/managing_countries/managing_provinces_of_country.feature [hunk 2]
```diff
@@ -25,6 +25,17 @@ Feature: Managing provinces of a country
...
         And this country should not have the "Northern Ireland" province
 
+    @ui @javascript @api
+    Scenario: Removing a province that is a zone member should not be possible
+        Given this country has the "Northern Ireland" province with "GB-NIR" code
+        And the store has a zone "All in GB" with code "GB"
+        And it has the "Northern Ireland" province member
+        When I want to edit this country
+        And I delete the "Northern Ireland" province of this country
+        And I save my changes
```
Comment (AdamKasp) [inline features/addressing/managing_countries/managing_provinces_of_country.feature:35]:
```suggestion
        And I try to save my changes
```

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+            $event->stopPropagation();
+        }
+    }
+
+    public function protectFromRemovingProvinceWithinCountry(GenericEvent $event): void
+    {
+        /** @var CountryInterface $country */
+        $country = $event->getSubject();
+        Assert::isInstanceOf($country, CountryInterface::class);
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:55]: Shouldn't we check to remove the subject that is `ProvinceInterface`?

File: features/addressing/managing_zones/deleting_zone.feature [hunk 4]
```diff
@@ -22,3 +25,9 @@ Feature: Deleting a zone
         When I delete zone named "South America"
         Then I should be notified that this zone cannot be deleted
         And I should still see the zone named "South America" in the list
+
+    @ui @api
+    Scenario: Deleting zone that is a zone member should not be possible
+        When I delete zone named "Central and Eastern Europe"
```
Comment (AdamKasp) [inline features/addressing/managing_zones/deleting_zone.feature:31]:
```suggestion
        When I try to delete zone named "Central and Eastern Europe"
```

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php (hunk 3 above)
Comment (coldic3) [reply src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:55]: The problem is we don't delete the Province directly but `orphanRemoval` configured in relation within a Country does it for us. That makes me unable to listen for an event `sylius.province.pre_delete` and I had to use `sylius.country.pre_update`.

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Kevin! :1st_place_medal:

**Pull request merged**